### ISSUE TITLE
Hardware-accelerated decoding

### ISSUE THREAD

**Issue by Kagami** [Kagami/mpv.js#5]
Right now CPU usage is rather high. Need to try to activate hwdec (at least on some platforms or add UI control) and check whether it's possible to improve performance in some way.

Might be useful: https://www.chromium.org/nativeclient/how-tos/3d-tips-and-best-practices

**Comment by ghost** [comment]
On Windows, it should work out of the box (as long as libmpv is linked to the correct ANGLE binary).

On Linux/Intel, you'd somehow need to thread through the Xlib handle or something similar. Linux/nvidia might work if chromium uses GLX.

https://github.com/mpv-player/mpv/blob/master/libmpv/opengl_cb.h#L112

**Comment by ghost** [comment]
When using mpv.js on Intel+Win7-x64 chipset the hardware acceleration seems to be disabled (CPU usage around 80% on when I run the example) whereas the standalone mpv player play the ToS file without loading the CPU more than 10%.

Any idea how to know why and what to do to solve this issue ?

**Comment by Kagami** [comment]
Do you have hwdec enabled in standalone mpv?

**Comment by ghost** [comment]
Yes I guess, in fact it is automatically detected by the standalone mpv (no need to provide any further command line arguments).
Whereas, when I try to force mpv.js to use hwdec (for instance provide it the argument vo=opengl or vo=direct3d) it craches or stays idle.

**Comment by ghost** [comment]
This seems much harder than I thought. It seems to go through some weird wrapper (I suspect we literally get WebGL, and the thing is probably designed to go through IPC). It also gets ES 2.0 only, which disables most useful advanced video rendering features.

I wonder if there's any way to sidestep the PPAPI wrappers, and get to native rendering APIs?

**Comment by TRPB** [comment]
Thanks for this great library. I can finally play HEVC videos in my electron app without transcoding them.

I think the issue with CPU usage is the output device. Hardware decode is working but the problem is that the output device is still software.

Currently `mpv.property('hwdec', 'vaapi');` seems to reduce CPU usage a little.

When setting `mpv.property('vo', 'vaapi');` or  `mpv.property('vo', 'gpu');`, the CPU usage is near-zero but the video is shown in its own window rather than in the browser. If it's somehow possible to capture the output when `vo` is set and display it in the app then I think this would be solved.

As a complete hack I wonder if it would be possible to open MPV's created window off the screen and  use WebRTC to pull the output into the electron browser window.


edit: `mpv.property('hwdec', 'vaapi-copy');` gives the best results on linux, assuming supported devices, etc. Not quite as low as `mpv` on its own but gives me less than half the CPU usage I get with `mpv.property('hwdec', 'vaapi');`
